src/main/java/com/minecolonies/structures/helpers/Structure.java

```diff
@@ -82,63 +92,361 @@
      * @param structureName name of the structure (at stored location).
      * @param settings      it's settings.
      */
-    public Structure(@Nullable final World world, final String structureName, final PlacementSettings settings)
+    public Structure(@Nullable final World world, String structureName, final PlacementSettings settings)
     {
-        InputStream inputstream = MinecraftServer.class.getResourceAsStream("/assets/" + Constants.MOD_ID + "/schematics/" + structureName + ".nbt");
-
         if (world == null || world.isRemote)
         {
             this.settings = settings;
             this.mc = Minecraft.getMinecraft();
         }
 
-        //Might be at a different location!
-        if (inputstream == null)
+        InputStream inputStream = null;
+        //Try the cache first
+        if (Structures.hasMD5(structureName))
         {
-            try
+            inputStream = Structure.getStream(Structures.SCHEMATICS_CACHE + '/' + Structures.getMD5(structureName));
+            if (inputStream != null)
             {
-                final File decorationFolder;
+                structureName = Structures.SCHEMATICS_CACHE + '/' + Structures.getMD5(structureName);
+            }
+        }
 
-                if (FMLCommonHandler.instance().getMinecraftServerInstance() == null)
-                {
-                    decorationFolder = new File(Minecraft.getMinecraft().mcDataDir, "minecolonies/");
-                }
-                else
-                {
-                    decorationFolder = new File(FMLCommonHandler.instance().getMinecraftServerInstance().getDataDirectory(), "minecolonies/");
-                }
-                if(decorationFolder.exists())
+        if (inputStream == null)
+        {
+            inputStream = Structure.getStream(structureName);
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER') Close this "InputStream". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2095)